Add typescript/#42

talesdsp opened the pull request.

talesdsp committed:
add deps + command

talesdsp committed:
add directory

talesdsp committed:
setup typescript

talesdsp committed:
remove js files

talesdsp committed:
add types for components

talesdsp committed:
move typings to types folder

talesdsp committed:
refactor

talesdsp commented:
Obs: I found that sometimes running tests with dist folder created throws an Emit skip error. Just warning in case it happens with you.

talesdsp commented on src/Block.tsx:
```diff
@@ -0,0 +1,373 @@
+      shadowRadius: elevation
+    },
+    space && { justifyContent: `space-${space}` },
+    card && { borderRadius: 12 },
```
Apparently SIZES.border is not declared, and I don't know what the value should be

hetmann replied:
We need to add a new value for SIZES.borderRadius

talesdsp commented on src/theme.ts:
```diff
@@ -0,0 +1,67 @@
+  info: "#4DA1FF"
+};
+
+export const SIZES: Sizes = {
```
The SIZES constant. It should be radius?

hetmann replied:
SIZES is an object with values for different usage.
For example: **radius** is used for **borderRadius**

talesdsp replied:
Yes! I should have been clear than that. What I mean is, the SIZES.radius is the correct value for the failing test below?

talesdsp commented on src/global.ts:
```diff
@@ -0,0 +1,157 @@
+  info: string;
+};
+
+export type Sizes = {
```
The SIZES type

talesdsp commented on src/__tests__/Block.test.tsx:
```diff
@@ -421,7 +421,8 @@ describe("<Block />", () => {
-    expect(style.borderRadius).toEqual(SIZES.border);
+    //!TODO: border value
+    expect(style.borderRadius).toEqual(SIZES.height);
   });
```
Needs fix too

talesdsp replied:
This test

hetmann commented on src/Button.tsx:
```diff
@@ -106,14 +108,12 @@ function Button(props) {
     outlined,
     flex,
     height,
-    borderWidth,
```
The last commit had the fix for `borderWidth` & `borderColor`

hetmann commented on src/Button.tsx:
```diff
@@ -186,43 +187,39 @@ function Button(props) {
-  const backgroundColor = StyleSheet.flatten(buttonStyles).backgroundColor;
-
   if (disabled) {
+    const backgroundColor = StyleSheet.flatten(buttonStyles).backgroundColor;
```
This can be extracted outside the if condition as it is used multiple times. See the last PR merge

hetmann reviewed (changes requested):
I notice that the latest changes for #33 where not included in your PR.

hetmann commented:
@taleco22 I would like to add ONLY the type definitions for components as requested in #42 
This will keep the library as light as possible, using a **.d.ts** file for the component/s.

talesdsp committed:
especify files in ts config

talesdsp committed:
add scripts and filter npm files

talesdsp committed:
delete files

talesdsp committed:
add components types

talesdsp committed:
move interfaces

talesdsp committed:
add type definition for components

talesdsp committed:
Merge branch 'master' of github.com:react-ui-kit/expo-ui-kit into add-typescript/#42

talesdsp commented:
I 've fixed what was missing from #33. Extracted the background. Removed the .tsx files. And kept just the type definitions now 😅👍 

Is there anything else?

hetmann commented on src/Button.js:
```diff
@@ -205,7 +209,7 @@ function Button(props) {
   }
 
-  const ButtonType = highlight
+  const touch = highlight
```
**touch** can be named **Touchable** or **ButtonInstance** and instead of **RenderButton** you can call it like this:
```js
<Touchable
  {...extraProps}
  disabled={disabled}
  activeOpacity={opacity}
  style={buttonStyles}>
  {children}
</Touchable>
```

hetmann commented:
> I 've fixed what was missing from #33. Extracted the background. Removed the .tsx files. And kept just the type definitions now 😅👍
> 
> Is there anything else?

@taleco22 Yes, I reviewed the code and I have left a comment on the Button component

talesdsp committed:
naming instance

msanvarov commented:
Hey @hetmann, love the work that is being done with this library, but I plead that you allow @talesdsp to continue with Typescript development. In my attempts to extend and even use this component library, I have come across some issues that can be solved if this library is ported back to Typescript. I'm willing to contribute along with @talesdsp to get it back to the state it was in before. I think declaration files are not enough for this type of ui library. Here is an example of another popular react native library built with Typescript: https://github.com/akveo/react-native-ui-kitten.

talesdsp commented:
@msanvarov 
Issues with usage? Can you be more specific?
Well, if the project turns out to typescript again mostly is already done up here. Won't be a pain to re-add it.

hetmann commented:
@talesdsp @msanvarov guys you can create a new branch named "ts" or "typescript" and use that in your projects, would that be helpful?

msanvarov commented:
> @msanvarov
> Issues with usage? Can you be more specific?
> Well, if the project turns out to typescript again mostly is already done up here. Won't be a pain to re-add it.

To begin with the utility helpers don't have types, so using them means I have to copy the test cases to understand the implementation. In terms of the implementation,  stuff like this  `const isArray = typeof padding === "object";` seems weird because to check if a variable is an array, the following can be used `Array.isArray(padding)`

I also don't understand why this is done:
```ts
export interface CardProps extends CardOptions, ThemeProps {
  [key: string]: any;
}
```
This defeats the purpose of typing if you are going to allow any key value to be accepted as a prop.

hetmann commented:
> > @msanvarov
> > Issues with usage? Can you be more specific?
> > Well, if the project turns out to typescript again mostly is already done up here. Won't be a pain to re-add it.
> 
> To begin with the utility helpers don't have types, so using them means I have to copy the test cases to understand the implementation. In terms of the implementation, stuff like this `const isArray = typeof padding === "object";` seems weird because to check if a variable is an array, the following can be used `Array.isArray(padding)`
> 
> I also don't understand why this is done:
> 
> ```ts
> export interface CardProps extends CardOptions, ThemeProps {
>   [key: string]: any;
> }
> ```
> 
> This defeats the purpose of typing if you are going to allow any key value to be accepted as a prop.

I was using the old style "typeof" checking for types :)
To understand the utils you can read the comments, I did my best to add an easy to understand usage.

msanvarov commented:
> > > @msanvarov
> > > Issues with usage? Can you be more specific?
> > > Well, if the project turns out to typescript again mostly is already done up here. Won't be a pain to re-add it.
> > 
> > 
> > To begin with the utility helpers don't have types, so using them means I have to copy the test cases to understand the implementation. In terms of the implementation, stuff like this `const isArray = typeof padding === "object";` seems weird because to check if a variable is an array, the following can be used `Array.isArray(padding)`
> > I also don't understand why this is done:
> > ```ts
> > export interface CardProps extends CardOptions, ThemeProps {
> >   [key: string]: any;
> > }
> > ```
> > 
> > 
> > This defeats the purpose of typing if you are going to allow any key value to be accepted as a prop.
> 
> I was using the old style "typeof" checking for types :)
> To understand the utils you can read the comments, I did my best to add an easy to understand usage.

I think it will be cool is get jsdoc integration for the examples you made then when we create a gitbook to detail using this library, it will be easier to showcase some use cases for it.

hetmann commented:
> > > > @msanvarov
> > > > Issues with usage? Can you be more specific?
> > > > Well, if the project turns out to typescript again mostly is already done up here. Won't be a pain to re-add it.
> > > 
> > > 
> > > To begin with the utility helpers don't have types, so using them means I have to copy the test cases to understand the implementation. In terms of the implementation, stuff like this `const isArray = typeof padding === "object";` seems weird because to check if a variable is an array, the following can be used `Array.isArray(padding)`
> > > I also don't understand why this is done:
> > > ```ts
> > > export interface CardProps extends CardOptions, ThemeProps {
> > >   [key: string]: any;
> > > }
> > > ```
> > > 
> > > 
> > > This defeats the purpose of typing if you are going to allow any key value to be accepted as a prop.
> > 
> > 
> > I was using the old style "typeof" checking for types :)
> > To understand the utils you can read the comments, I did my best to add an easy to understand usage.
> 
> I think it will be cool is get jsdoc integration for the examples you made then when we create a gitbook to detail using this library, it will be easier to showcase some use cases for it.

I have in plan to do an actual app for this UI kit :) see #23 
Also the jsdoc or gitbook would be nice to have
